crypto/aes/asm/aes-cfb-avx512.pl

```diff
@@ -0,0 +1,922 @@
+
+    test $len,$len                    # return early if $len==0, unlikely to occur
+    jz .Laes_cfb128_vaes_dec_done
+___
+
+$code.=<<___ if($win64);
+    sub \$0x10,%rsp
+.cfi_adjust_cfa_offset 16
+    vmovdqu %xmm6,(%rsp)              # xmm6 needs to be maintained for Windows
+___
+
+$code.=<<___;
+    test $num,$num                    # check if the first block is partial
+    jz .Laes_cfb128_dec_mid           # if not, jump to processing full blocks
+
+###########################################################
+# first partial block pre-processing
+###########################################################
+
+    mov $key_original,$key_backup     # make room for variable shl with cl
+
+    mov \$0x10,$left                  # first block is partial
+    sub $num,$left                    # calculate how many bytes $left to process in the block
+    cmp $len,$left                    #
+    cmova $len,$left                  # $left = min(16-$num,$len)
+
+    mov \$1,$mask                     # build a mask with the least significant $left bits set
+    shlq %cl,$mask                    # $left is left shift counter
+    dec $mask                         # $mask is 2^$left-1
+    kmovq $mask,%k1
+
+    mov $num,%rax                     # keep in-out num in %al
+    add $left,%rax                    # advance $num
+    and \$0x0F,%al                    # wrap-around $num in a 16-byte block
+
+    leaq ($num,$ivp),%r11             # process $left iv bytes
+    vmovdqu8 (%r11),%xmm0
+    vmovdqu8 ($inp),%xmm1             # process $left input bytes
+    vpxor %xmm0,%xmm1,%xmm2           # CipherFeedBack XOR
+    vmovdqu8 %xmm2,($out){%k1}        # write $left output bytes
+    vmovdqu8 %xmm1,(%r11){%k1}        # blend $left input bytes into iv
+
+    add $left,$inp                    # advance pointers
+    add $left,$out
+    sub $left,$len
+    jz .Laes_cfb128_dec_zero_pre      # return early if no AES encryption required
+
+    mov $key_backup,$key_original     # restore "key_original" as arg3
+
+.Laes_cfb128_dec_mid:
+___
+
+    &load_aes_key_schedule_4x();
+
+$code.=<<___;
+###########################################################
+# inner full blocks processing
+###########################################################
+
+    # $temp_4x is "iv | iv | iv | iv"
+    vbroadcasti32x4 ($ivp),$temp_4x             # load iv
+
+    cmp \$0x100,$len                            # are there 16 ciphertext blocks left (2048 bits) ?
+    jb .Laes_cfb128_dec_check_4x
+
+###########################################################
+# decrypt groups of 16 128-bit blocks in parallel
+# behaves as 16x loop unroll
+###########################################################
+
+.balign 32
+.Loop_aes_cfb128_dec_mid_16x:
+    sub \$0x100,$len
+
+    # load 16 ciphertext blocks
+
+    # $cipher_4x  is "ciphertext  0 | ciphertext  1 | ciphertext  2 | ciphertext  3"
+    vmovdqu32 ($inp),$cipher_4x
```

Comment:
Could you use vmovdqu64 instead when there is no op mask? (maybe I am misreading what the difference is)

Reply:
For unmasked operations the performance and register values will be identical. The difference between the two is about expressing intent.

There are several ways to perform 512-bit [unaligned loads](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=mm512_loadu):
- `vmovdqu64` / `_mm512_loadu_epi64` would mean loading 8 packed 64-bit integers
- `vmovdqu32` / `_mm512_loadu_si512` would mean loading a continuous chunk of 512 bits

I believe there is no instruction/intrinsic for loading 4 `si128` chunks so I chose the latter.